File 'encall.sh' Gets a Rework

terminalforlife opened the pull request:
Please read the commit message _before_ merging. ;)

Commit:
Clean up the code by making it more concise

Please be _sure_ to check this _before_ merging, in-case I've missed
something. I checked bits and pieces, and some I assume will be fine, -
as I've done them many times. You never, though.

The general gist, is that the `Do()` function serves to avoid code
repetition, and, as your comment suggested you wanted, it's now using
`find`.

However, to keep the output on the one line, being able to see each
file as its being processed, is done here thanks to a few things:

  * The `setterm` command disables line-wrapping, avoiding new lines.

  * The `-printf` flag's `\r` clears from cursor position to column 0.

  * The `echo` creates the final new line character.

Line wrapping _should_ be re-enabled before `Do()` finishes.

The whole _default password if none is given_ thing is, at least in
this case, better done using `:-`, as shown, which the shell will end
up seeing it as _if not set, then instead expand to this_. No
assignment takes place, by the way. Also, `:+` is the opposite.

Replaced the main `if` statement with a `case` statement, to condense.

Commit:
Minor changes (bringing `echo` back)

bing0o commented:
hello @terminalforlife ,
first thanks for all your modifications you're helping me a lot to understand bash scripting better :) ,
in this one it didn't work for me , there's a different in the find command when the option e (for encryption) and d (for decryption), the first (e) we collect all the files in the given path except of the files that ends with .hacklab, which is already encrypted, like : 
```bash
find . | grep -v "\.hacklab$" 
```
> i don't know how to use grep within -exec here instead of (| pipe ) 

The second option (d) we collect only the files that ends with .hacklab to be decrypted, like:
```bash
find . -name ".hacklab"
```
also i didn't use `-maxdepth` here cuz i thought it's better to encrypt all the files in the current directory and sub directories too, but let's keep `-maxdepth` option it looks better.  

[Crypto.py](https://github.com/bing0o/Python-Scripts/blob/master/crypto.py)  script takes 3 options, -e/-d the file to be encrypted or decrypted , -p the password, -x to delete the original file. 
also i don't understand why you're giving `-not` in e, and `' '` in d here !?.

```bash
case $1 in
	e)
		Do -not -e ;;
	d)
		Do ' ' -d ;;
```
i think we only need the `e`, `d` and the `password` options here.

i'll try to edit your script and back to you.

bing0o commented:
@all-contributors please add @terminalforlife

bing0o commented:
@all-contributors please add @terminalforlife  for code

ghost commented:
You're welcome. lol :)

Regarding the `' '` and `'-not'` part, it's because I wanted the first argument to the function to be taken as a whitespace, in order to prevent the 2nd argument from becoming the first, since no first would have then been detected without the quoted space.

```bash
setterm -linewrap off
find -maxdepth 1 \( -type f $1 -name '*.hacklab'\
	-exec crypto $1 {} -p ${2:-P4ssw@rD} -x 1>/dev/null \;\
	-exec tput el \; \) -printf "%p\r"
setterm -linewrap on
printf "[*] Done!\n"
```

The code here (from the `Do()` function) _should_ non-recursively list all files found within the current directory, but omitting files which _don't_ end in '.hacklab'. If I understood your explanation (thank you, by the way) correctly, this should give you the result you're after. However, as you've said, it's not working. What does happen? Any chance in posting some output, if it's not too sensitive?

bing0o commented:
hello @terminalforlife 
I think i figured it out, this is the changes that i made:
first in this line 
```bash
-exec crypto $1 {} -p ${2:-P4ssw@rD} -x 1>/dev/null \;\
```
the first option `$1` is `-not` or `' '` and this is an unknown options for crypto.py script
so i changed it to this:
```bash
-exec crypto $2 {} -p ${3:-P4ssw@rD} -x 1>/dev/null \;\
```
when the `$2` is `-e` or `-d` and the `$3` is the password 

and the case will be :
```bash
case $1 in
        e)
                Do -not -e $2;;
        d)
                Do ' ' -d $2;;
        ''|*)
                echo "Usage: encall <OPTIONS (e/d)> <PASSWORD, default (P4ssw@rD)>"
                exit 1 ;;
esac
```
but there's no output , it should shows the processed file !

thank for your time!.

ghost commented:
Ooooh! I'm sorry, I must have had a brain-fart and gotten the arguments backwards. -_- Looks like you did indeed fix it, and well done for figure it out!

I'm going to revisit the file I edited today, because I'm not too happy with what I changed, considering your feedback. So grizzly bear with me, got a bazillion projects in the air. lol

bing0o commented:
@terminalforlife thanks for your help man, have a good day and happy coding :D 
give me your twitter or discord account if you can !.

ghost commented:
:) You're more than welcome, and I don't have either accounts, AFAIK. I'm on [YouTube](https://www.youtube.com/channel/UCfp-lNJy4QkIGnaEE6NtDSg) as well though. Also, you can always E-Mail me over at: terminalforlife@yahoo.com I don't check the E-Mails nearly as often as I do YouTube and GitHub though. I'm on GitHub pretty much daily.

I'm also on [Reddit](https://www.reddit.com/user/LearnLinux_onYT), which I check from time to time, mainly on `r/bash`, `r/shell`, and similar subreddits.

bing0o commented:
OK thanks again :D 
I'm already following you on youtube, you're doing great on it :)    

Next time :+1:

ghost commented:
Hi. Sorry for getting back to you so late. Been busier than I expected.

I was just looking at the code I sent you, and it looks like I missed the password part of the argument processing. I'm going to just sort this. I feel bad for wasting your time, but I also appreciate that this has been a learning moment for us both. lol I think you'll like what I have cooking though...

Commit:
Revise and overhaul `encall.sh`

I checked out to e41f7317a9ad8d5f27095a6908d7cf888 commit, made by
@bing0o on November the 23rd, then just redid the file. I went all out
this time. Hopefully I didn't overlook something.

Bug Fixes:

  * Should now address the _All-Elusive Argument of 2019_ issue.

  * Caught error when the user enters no password but uses its flag.

  * No longer parses `ls`, thus avoiding lots of unwanted effects.

Additions:

  * Reworked the header, and fancied it up a bit.

  * Added `CurVer` and `Progrm` variables.

  * Added `Err()` for consistent and presentable error messages.

  * Added an informative `Usage()` function, akin to my own programs.

  * Added each of us (your E-Mail pending, if applicable) to `Usage()`.

  * Reworked the description, especially to be in line with changes.

  * The synopsis or syntax is now more traditional, and easier to read.

  * Preferred the standard GNU long and short flag styling: `--`, `-`

  * Added long options for the old arguments.

  * Added `--help|-h|-?` and `--version|-v` options.

  * Added a `SITE:` link to both of your related repositories.

  * Now iterates over whichever files are specified by the user.

  * Tweaked the `Processing:` message to be more situational.

Code:

  * Prefer `printf` over `echo`, for portability and functionality.

  * Written and converted to POSIX Bourne for excellent portability.

terminalforlife commented on encall.sh:
```diff
@@ -1,29 +1,95 @@
-#!/bin/bash
+#!/bin/sh
+#cito M:755 O:0 G:0 T:/usr/local/bin/crypto
```
Feel free to remove this. Admittedly, it's more of a plug. :P But hey, it is pretty convenient. lol

bing0o replied:
hi @terminalforlife , you're back with new great update man, it's a completely different from the previous version , I tested it everything looks great except of that if you run the script without any argument it should show the usage (help) massage, but instead it will try to execute crypto without `-e` or `-d` and an error will appear !.
Also when i use `-d` it will try to decrypt even the files that doesn't end with `.hacklab` (not an encrypted files).

I really appreciate your time and this work , it's great chance for me to learn from you :)

terminalforlife commented on encall.sh:
```diff
@@ -1,29 +1,95 @@
 #
-# script to encrypt all the files in a specific directory, using crypto.py python script.
+#   wget https://raw.githubusercontent.com/bing0o/Python-Scripts/master/crypto.py
+#   chmod 755 crypto.py; chown 0:0 crypto.py
```
Forgot to mention in the commit message, that I made this a _whole_ lot safer.

Commit:
Merge branch 'master' of https://github.com/bing0o/bash_scripting

Commit:
Address issues mentioned by bing0o in PR

I've also added a little dependency check for crypto, in-case the user
doesn't have it.

ghost commented:
Yeah, sorry -- I forgot to catch the error when the user doesn't specify any files. Hopefully I've addressed the issues you mentioned. Let me know how it goes.

bing0o commented:
the decryption issue is fixed but still try to run crypto instead of showing the help message when i run the script without arguments .

ghost commented:
That's something. Can you post the output you see?

bing0o commented:
this is the output:
![encallissue](https://user-images.githubusercontent.com/25457545/70816309-7fee9780-1dcf-11ea-9d81-864a83558d3d.png)

ghost commented:
Oooh, you're not even using the version from my fork lol That's the old usage info. It should look like this:

![Screenshot_2019-12-13_16:48:52](https://user-images.githubusercontent.com/31768530/70816977-7ca3dd80-1dc8-11ea-98e5-97154adfbbe4.jpg)

Maybe clone my fork in a different place altogether, then try it. I just tested it but just had the crypto stuff echoed, and it looks to have worked okay.

Commit:
Address no arguments exiting with zero status

bing0o commented:
nope i used the last version, and in your output you used `-h` but i mean what if i just run the script `./encall.sh`, it's not gonna print the help message.

Commit:
Address issue caused by no arguments at all

Not specifying any files is undesired, so also ouput the usage, with an
exit status of 1.

ghost commented:
There we go. Sorry. lol I'm having a lot of blonde moments, lately. -_-

bing0o commented:
you fixed it, thanks a lot :)
but now if i use `-e` or `-d` its just print the help message and exit, also it doesn't use the password that provided by `-p` option to encrypt or decrypt the files, instead it uses the default one :(

sorry i took your time for this!, i'll try to fix this one :)

ghost commented:
It's okay! I love helping. I'm sorry we keep having problems here. lol

Commit:
Dependency notice should take priority

ghost commented:
Oh wait, that's by design (regarding the `-e` or `-d` options) since they require a file. I think it used to just work on everything, but now you can actually specify a file, or multiple files. For example:

```sh
# Iterates over each file in the CWD, dot or not.
encall -e -p password * .*

# Various files in various places.
encall -d -p password file_1 /file/path/here.txt /and/so/on.tar
```

This kind of argument processing doesn't support, for example `-ep password`, unfortunately. Programming that in as well would be a huge pain in the ass.

I can't see why the password isn't working, though. :\

bing0o commented:
the password is working know , but the files even if i give it one file it's gonna take all the file for the process.
but it's okay, i think we are ready to merge this pull request :)
